contracts/overseer/src/contract.rs

```diff
@@ -287,6 +378,90 @@ pub fn update_whitelist(
     ]))
 }
 
+fn update_deposit_rate(deps: DepsMut, env: Env) -> StdResult<()> {
+    let dynrate_config: DynrateConfig = read_dynrate_config(deps.storage)?;
+    let dynrate_state: DynrateState = read_dynrate_state(deps.storage)?;
+    let mut config: Config = read_config(deps.storage)?;
+
+    // check whether its time to re-evaluate rate
+    if env.block.height >= dynrate_state.last_executed_height + dynrate_config.dyn_rate_epoch {
+        // retrieve interest buffer
+        let interest_buffer = query_balance(
+            deps.as_ref(),
+            env.contract.address.clone(),
+            config.stable_denom.to_string(),
+        )?;
+        // convert block rate into yearly rate
+        let blocks_per_year = Decimal256::from_ratio(Uint256::from(BLOCKS_PER_YEAR), 1);
+        let current_rate = config.threshold_deposit_rate * blocks_per_year;
+
+        let yield_reserve = Decimal256::from_uint256(interest_buffer);
+        let mut yr_went_up = yield_reserve > dynrate_state.prev_yield_reserve;
+
+        // amount yield reserve changed in notional terms
+        let yield_reserve_delta = if yr_went_up {
+            yield_reserve - dynrate_state.prev_yield_reserve
+        } else {
+            dynrate_state.prev_yield_reserve - yield_reserve
+        };
+
+        // amount yield reserve changed in percentage terms
+        // if the prev yield reserve was zero; assume either a 100% decrease
+        // or a 100% increase, but this should be very rare
+        let mut yield_reserve_change = if dynrate_state.prev_yield_reserve.is_zero() {
+            Decimal256::one()
+        } else {
+            yield_reserve_delta / dynrate_state.prev_yield_reserve
+        };
+
+        // decreases the yield reserve change by dyn_rate_yr_increase_expectation
+        // (assume (yr_went_up, yield_reserve_change) is one signed integer, this just subtracts
+        // that integer by dynrate_config.dyn_rate_yr_increase_expectation)
+        let increase_expectation = dynrate_config.dyn_rate_yr_increase_expectation;
+        yield_reserve_change = if !yr_went_up {
+            yield_reserve_change + increase_expectation
+        } else if yield_reserve_change > increase_expectation {
+            yield_reserve_change - increase_expectation
+        } else {
+            yr_went_up = !yr_went_up;
+            increase_expectation - yield_reserve_change
+        };
+
+        yield_reserve_change = min(yield_reserve_change, dynrate_config.dyn_rate_maxchange);
+
+        let mut new_rate = if yr_went_up {
+            current_rate + yield_reserve_change
+        } else if current_rate > yield_reserve_change {
+            current_rate - yield_reserve_change
+        } else {
+            Decimal256::zero()
+        };
+
+        // convert from yearly rate to block rate
+        new_rate = new_rate / blocks_per_year;
```

Comment:
Might be cleaner to just give this a name like `new_per_block_rate` so its easier to follow.